ocis-bin: init at 5.0.0

bhankas opened the pull request:
## Description of changes

Adds binary package for Owncloud Infinite Scale Stack. Fixes #230190.

There are currently issues building ocis for Nix independently, leaving us with options to fetch the prebuilt binary or the prebuilt container. 

A rudimentary NixOS module is also included.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Commit:
ocis-bin: init at 5.0.0

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/1016/112

MaxwellDupre reviewed (approved):
why do you need an rc build? What happens to this package when ocis goes to release?

bhankas commented:
> why do you need an rc build? What happens to this package when ocis goes to release?

5.0.0 has some significant changes and deprecations (see [release notes](https://doc.owncloud.com/ocis_release_notes.html#infinite-scale-5-0-0)). Since the package is just starting out in nixpkgs, I just went with rc. There is still couple months till next NixOS release, so still plenty of time.

All that said, I wouldn't mind changing it to stable release if you think that would be safer call.

bhankas commented:
@MaxwellDupre updated to have 4.0.6 stable tag instead.

Commit:
nixos/ocis: init at 5.0.0

Co-authored-by: Casey Link <unnamedrambler@gmail.com>

danth commented on pkgs/by-name/oc/ocis/package.nix:
```diff
@@ -0,0 +1,47 @@
+
+  meta = with lib; {
+    description = "ownCloud Infinite Scale Stack ";
+    homepage = "ownCloud Infinite Scale Stack ";
```
```suggestion
    homepage = "https://owncloud.dev/ocis/";
```

danth commented on pkgs/by-name/oc/ocis/package.nix:
```diff
@@ -0,0 +1,47 @@
+    changelog = "https://github.com/owncloud/ocis/releases/tag/v${version}";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ payas ];
+  };
```
Consider setting `sourceProvenance` to signify that this is using a prebuilt binary.

```suggestion
    sourceProvenance = [ sourceTypes.binaryNativeCode ];
  };
```

danth commented on pkgs/by-name/oc/ocis/package.nix:
```diff
@@ -0,0 +1,47 @@
+      else if stdenv.isLinux && stdenv.isx86_64 then
+        "c33e2055359d98c36dfa953b56a7a556d911169db27f0f829c9eb41cf37ff615"
+      else
+        "";
```
What happens if this value is reached?

Consider using `builtins.throw` to create a meaningful error message.

danth commented on nixos/modules/services/web-apps/ocis.nix:
```diff
@@ -0,0 +1,123 @@
+{
+  options = {
+    services.ocis = {
+      enable = lib.mkEnableOption "Enable ownCloud Infinite Scale Stack.";
```
"Whether to enable" is added automatically, so this should only contain the name.

```suggestion
      enable = lib.mkEnableOption "ownCloud Infinite Scale";
```

danth commented:
Feel free to add me as a maintainer.

bhankas commented:
@danth  thanks for review, addressed comments and added you as maintainer.

bhankas commented:
If 5.0 gets released before 24.05 freeze, we can update fairly easily, unstable is called that for good reason anyway. Stable is better in case there are more rcs.

danth commented:
Looks like 5.0 just got released :sweat_smile: 

https://github.com/owncloud/ocis/releases/tag/v5.0.0

SomeoneSerge commented on pkgs/by-name/oc/ocis/package.nix:
My preference would be to reserve `pkgs.ocis` for the source build and name this `ocis-bin` instead. That said, some would probably disagree (e.g. our `tor-browser` is misleadingly a binaryNativeCode package as well)

danth replied:
Having this as `pkgs.ocis` would mean we can easily replace it with the source package once the issues building it are resolved.

This depends on whether we would want to keep both `pkgs.ocis` and `pkgs.ocis-bin`, or only maintain one.

bhankas commented:
@danth nice catch, updated to 5.0.0.

@SomeoneSerge good point, addressed.

bhankas commented:
One question, since this is using the binary, and that appears to be under [EULA](https://github.com/owncloud/ocis/releases/download/v5.0.0/End-User-License-Agreement-for-ownCloud-Infinite-Scale.pdf), should the license be `agpl3Only` or should it be marked unfree?

danth commented:
It may be useful to add a comment for anyone who comes across this in the future and wonders why it's marked unfree.

danth commented:
A comment directly in the file would avoid the need to search through history to find it, but that's not a major issue.

bhankas commented:
Added a comment

Ramblurr commented on nixos/modules/services/web-apps/ocis.nix:
```diff
@@ -0,0 +1,114 @@
+        '';
+      };
+
+      settings = lib.mkOption {
```
This `settings` option is never written to a place where the user-supplied settings could actually be picked up by OCIS. It should probably be written as an environmentfile in the systemd service, no?

bhankas replied:
yea, it was somehow always a TODO item to use it. Now it gets written as environmentFile, prepended by the other environmentFile user may want and set

Ramblurr commented on nixos/modules/services/web-apps/ocis.nix:
```diff
@@ -0,0 +1,114 @@
+      wantedBy = [ "multi-user.target" ];
+      environment = {
+        PROXY_HTTP_ADDR = "${cfg.settings.address}:${toString cfg.settings.port}";
+        OCIS_INSECURE = "false";
+        OCIS_LOG_LEVEL = "error";
```
These should be overrideable by the user's settings.

bhankas replied:
Done

bhankas commented on nixos/modules/services/web-apps/ocis.nix:
```diff
@@ -0,0 +1,114 @@
+            stateDir = lib.mkOption {
+              default = "/var/lib/ocis";
+              type = types.str;
+              description = "Workout-tracker data directory.";
```
oops, fixed

bhankas commented on nixos/modules/services/web-apps/ocis.nix:
```diff
@@ -0,0 +1,114 @@
+        Type = "simple";
+        ExecStart = "${lib.Exe cfg.package} server";
+        WorkingDirectory = cfg.stateDir;
+        DynamicUser = true;
```
TIL, thanks! Created static and use overridable user+group, and used that in service config

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1555

Ramblurr commented:
@bhankas I think the license change should be squashed into the `ocis-bin: init at 5.0.0` commit.

The nixos module should be added to `nixos/modules/module-list.nix` and referenced in `nixos/doc/manual/release-notes/rl-2405.section.md` under "New Services"

bhankas commented:
> @bhankas I think the license change should be squashed into the `ocis-bin: init at 5.0.0` commit.

Done.

> The nixos module should be added to `nixos/modules/module-list.nix` and referenced in `nixos/doc/manual/release-notes/rl-2405.section.md` under "New Services"

Done and Done!

bhankas commented:
I  have no idea why manual building is failing. Its very likely to be something really dumb, but help appreciated

danth commented on nixos/modules/services/web-apps/ocis.nix:
```diff
@@ -0,0 +1,131 @@
+{
+  config,
+  lib,
+  pkgs,
+}:
```
Adding `...` here should fix it.

```suggestion
{ config, lib, pkgs, ... }:
```

bhankas replied:
Nope, still borked :(

Ramblurr commented:
Running locally (with the `...` fix) I am getting:

```
       Failed assertions:
       - Exactly one of users.users.ocis.isSystemUser and users.users.ocis.isNormalUser must be set.
       ```

bhankas commented:
Casey Link ***@***.***> writes:

> Running locally (with the `...` fix) I am getting:
>
> ```
>        Failed assertions:
>        - Exactly one of users.users.ocis.isSystemUser and users.users.ocis.isNormalUser must be set.
>        ```

Thanks! I set `isSystemUser=true` but doesn't appear to work. Doesn't
help the fact that my machine is slow and network is even slower and
unreliable to run `nixpkgs-review`.

Its a bit of work, but I am happy to keep fixing anything that you or
others may find, this is among my very first module PRs so having to fix
fair bit.
--

bhankas commented:
Hey, it worked! It was indeed dumb :) I was using old nested option link, just had to update it.

@danth , @Ramblurr thanks for putting up, feel free to take another look.

Ramblurr commented on nixos/modules/services/web-apps/ocis.nix:
```diff
@@ -0,0 +1,133 @@
+        };
+        serviceConfig = {
+          Type = "simple";
+          ExecStart = "${lib.Exe cfg.package} server";
```
`lib.Exe` doesn't exist

Ramblurr replied:
Also this line should probably be something like: `ExecStart = "${cfg.package}/bin/ocis server";
`

danth replied:
I think this was a typo of `lib.getExe`.

`lib.getExe cfg.package` expands to `${cfg.package}/bin/${cfg.package.meta.mainProgram}`, and in this case `meta.mainProgram` should be set to `"ocis"`.

bhankas replied:
Lack of `meta.mainProgram` is also heavily discouraged in nixos module tests. So I've added it now.

Typo has been fixed.

Ramblurr commented on nixos/modules/services/web-apps/ocis.nix:
```diff
@@ -0,0 +1,133 @@
+  };
+
+  config = lib.mkIf cfg.enable {
+    users.users.${cfg.user} = {
```
It was better before you changed it to unconditionally create the user. The convention is that if the caller changes the default user, then they're responsible for creating it.

Use something like:
```nix
    users.users.${defaultUser} = lib.mkIf (cfg.user == defaultUser) {
      group = cfg.group;
      home = cfg.dataDir;
      isSystemUser = true;
      createHome = true;
      description = "ownCloud Infinite Scale daemon user";
    };

    users.groups = lib.mkIf (cfg.group == defaultGroup) { ${defaultGroup} = { }; };
```

bhankas replied:
Thanks, done!

Ramblurr commented on nixos/modules/services/web-apps/ocis.nix:
```diff
@@ -0,0 +1,133 @@
+          ExecStart = "${lib.Exe cfg.package} server";
+          WorkingDirectory = cfg.stateDir;
+          User = cfg.user;
+          Group = cfg.Group;
```
capitalization typo

bhankas replied:
Thanks, fixed!

Ramblurr commented on nixos/modules/services/web-apps/ocis.nix:
```diff
@@ -0,0 +1,133 @@
+        description = ''
+          The user to run oCIS as.
+          By default, a user named `${defaultUser}` will be created whose home
+          directory is [dataDir](#opt-services.ocis.stateDir).
```
no `cfg.dataDir` option is defined

bhankas replied:
oops, fixed!

Ramblurr commented on nixos/modules/services/web-apps/ocis.nix:
```diff
@@ -0,0 +1,133 @@
+  config = lib.mkIf cfg.enable {
+    users.users.${cfg.user} = {
+      group = cfg.group;
+      home = cfg.dataDir;
```
no `cfg.dataDir` option is defined

bhankas replied:
oops, fixed!

Ramblurr commented on nixos/modules/services/web-apps/ocis.nix:
```diff
@@ -0,0 +1,133 @@
+    users.groups.${cfg.group} = { };
+
+    systemd = {
+      packages = [ pkgs.ocis-bin ];
```
This line is not necessary and incorrect. Referencing `${cfg.package}` in the ExecStart is sufficient.

danth replied:
To clarify, `systemd.packages` would only be necessary if the package contained a file such as `share/systemd/system/ocis.service`. This is generally used when a systemd service definition was already provided by the upstream developers, and we want to reuse it without converting it into Nix.

bhankas replied:
TIL!
Also, removed.

bhankas commented:
@Ramblurr thanks a ton for pointing to the tests, it was indeed super helpful in ironing out the kinks. I've added a test, but since this is unfree package it won't be run by CI nor hydra. It should still be valuable for future module maintenance.

@danth I've learned crapton of systemd-units and nixpkgs from your reviews, thanks for the pointers!

Anyway, I think we are good for next round. Hit me with your finest :)

P.S. I'll probably add config options for enabling/disabling every service in future, because right now we are running the so called `fullstack` server with every service supported by ocis. I think it is fine for now, because settings example gives all the values that are necessary, but further useability and/or customizability for distributed deployment will certainly help. That can wait though, but let me know what you think.

Ramblurr commented:
Awesome. I'll give it a spin tomorrow. For my use case the full stack mode is fine. It's really nice OCIS has that feature.

bhankas commented:
> Awesome. I'll give it a spin tomorrow. For my use case the full stack mode is fine. It's really nice OCIS has that feature.

One problem with the fullstack mode is it starts 33 (!) services right now, that block 33 ports, and the user does not get any control over them. That needs to be addressed.

I need to up my Nix-game. There are [so many services](https://doc.owncloud.com/ocis/next/deployment/services/services.html), and I'm sure new ones will get added over time, that manually adding config options for them all is going to be painful, not to mention maintenance and debugging.

I'd like to just make a list and generate all the systemd service configs for each of them, but again, my Nix-fu is pretty weak ATM. If you know any other module that does something (autogenerate systemd service config from some list), I'll be happy to work on it over coming weeks.

danth commented:
The basic structure could be something like

```nix
systemd.services = lib.genAttrs
  [ "service-one" "service-two" ]
  (name: {
    serviceConfig = { };
  });
```

bhankas commented:
Daniel Thwaites ***@***.***> writes:

> The basic structure could be something like
>
> ```nix
> systemd.services = lib.genAttrs
>   [ "service-one" "service-two" ]
>   (name: {
>     serviceConfig = { };
>   });
> ```

Thanks, this will help me get started.

Just to reiterate, that will take time, and I'd prefer it to not hold
this PR back. If it is a blocker, I'd rather close the PR and come back
whenever I'm able to work out service segregation.
--

danth commented:
Personally I don't think that should be a blocker for this PR.

bhankas commented:
Thanks, then I'll wait for further review/approval.

On Thu, 28 Mar, 2024, 15:07 Daniel Thwaites, ***@***.***>
wrote:

> Personally I don't think that should be a blocker for this PR.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/296679#issuecomment-2024774034>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFZBOUNDSN3NFUAXB7OA3E3Y2PJDTAVCNFSM6AAAAABE2JX6HKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMRUG43TIMBTGQ>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

Ramblurr commented on nixos/modules/services/web-apps/ocis.nix:
```diff
@@ -0,0 +1,154 @@
+          OCIS_CONFIG_DIR = "${cfg.stateDir}/config";
+          OCIS_BASE_DATA_PATH = cfg.stateDir;
+        } // cfg.settings;
+        serviceConfig = {
```
It would be nice to add the security and sandboxing features of systemd, which is becoming the norm for new systemd services in nix to increase security + reduce attack surface. See my other comment about that.

bhankas replied:
yeah, I'll need to read more on this. See other reply, it might take me some time, so if you have something more ready, we can use that :)

Ramblurr reviewed (changes requested):
Nice work @bhankas I think we are almost there.

I made a few comments to the PR.. but I also have some further suggestions. I coded them up and pushed them to [my nixpkgs ocis branch](https://github.com/Ramblurr/nixpkgs/commit/70b2ee9c1274c5a123c13687b3b883678403c2c6)

In particular the changes I'm suggesting:

* In the nixos test file: use headless firefox to sign in using a username + password and loading the main ocis dashboard screen. This better verifies that ocis is actually running and working
  *  I'm so amazed how fast and simple the nixos integration tests are.. it's really awesome
* Adds a `ocis-init.service` to run `ocis init`, optionally with a `cfg.adminPasswordFile` to set the admin password (otherwise it is randomly generated and spit out into the journal)
* Adds systemd hardening options to the systemd services
* Fixes the other issues I mentioned in the PR review

Finally: I hope I'm not stepping on your toes or anything by suggesting such a big change like this. Feel free to copy the changes into this PR, no need to credit me.

I'm also happy to be added as a maintainer if you want. 

My plan this week was to submit a nixos module for ocis, so it's nice we are able to collaborate on it together.

bhankas commented:
Casey Link ***@***.***> writes:

> @Ramblurr requested changes on this pull request.
>
> Nice work @bhankas I think we are almost there.
>
> I made a few comments to the PR.. but I also have some further suggestions. I
> coded them up and pushed them to [my nixpkgs ocis
> branch](https://github.com/Ramblurr/nixpkgs/commit/70b2ee9c1274c5a123c13687b3b883678403c2c6)

Thanks! I fixed the small ones quickly enough, but haven't yet checked
out your branch.

> In particular the changes I'm suggesting:
>
> * In the nixos test file: use headless firefox to verify that logging in
> actually works. This better verifies that ocis is actually running and working
>   *  I'm so amazed how fast and simple the nixos integration tests are.. it's really awesome

Yes! Always wondered why people went on about NixOS tests, but writing
one and actually witnessing it in action is quite something!

> * Adds a `ocis-init.service` to run `ocis init`, optionally with a
> `cfg.adminPasswordFile` to set the admin password (otherwise it is randomly
> generated and spit out into the journal)
> * Adds systemd hardening options to the systemd services
> * Fixes the other issues I mentioned in the PR review

After having to go through the service docs and actually getting the
bloody thing to start up has me unsure on `ocis-init.service`. We
want users to be configuring their ADMIN_PASSWORD declaratively and put
it in their config using something like sops-nix/agenix. Running an
imperative operation that creates a config, which can easily get
lost/moved/deleted in sea of other files (e.g. impermanence) is undesired,
especially for a service that is intended to not lose users' files,
that'd be ironic :)

> Finally: I hope I'm not stepping on your toes or anything by suggesting such a
> big change like this. Feel free to copy the changes into this PR, no need to
> credit me.

No problem, always nice to have someone to work with than otherwise :)

I will probably be able to take a look at your branch later today, but the weekend
will probably be busy for me to make big changes. If you feel your
branch is further along, I don't mind closing this PR and going with
yours, no feelings hurt :)

> I'm also happy to be added as a maintainer if you want.

Thanks, I've added you in maintainers.

> My plan this week was to submit a nixos module for ocis, so it's nice we are
> able to collaborate on it together.

Definitely. Let me know what you think.

--

bhankas commented on nixos/modules/services/web-apps/ocis.nix:
```diff
@@ -0,0 +1,154 @@
+        } // cfg.settings;
+        serviceConfig = {
+          Type = "simple";
+          ExecStart = "${lib.getExe cfg.package} server web";
```
oops, I think this was a leftover from a debug session, fixed.

bhankas replied:
Not sure on manual or automated `ocis init`, see the other, longer reply.

bhankas commented on nixos/modules/services/web-apps/ocis.nix:
```diff
@@ -0,0 +1,154 @@
+          OCIS_SERVICE_ACCOUNT_ID = "foo";
+          OCIS_SERVICE_ACCOUNT_SECRET = "foo";
+          OCIS_SYSTEM_USER_API_KEY = "foo";
+          STORAGE_USERS_MOUNT_ID = "123";
```
Mentioned in settings docstring.

bhankas commented on nixos/modules/services/web-apps/ocis.nix:
```diff
@@ -0,0 +1,154 @@
+        description = "Web interface port.";
+      };
+
+      domain = lib.mkOption {
```
done

bhankas commented on nixos/modules/services/web-apps/ocis.nix:
```diff
@@ -0,0 +1,154 @@
+
+      domain = lib.mkOption {
+        type = types.str;
+        default = "http://localhost:9200";
```
done

bhankas commented:
Casey Link ***@***.***> writes:

>> After having to go through the service docs and actually getting the bloody thing to start up has me unsure on `ocis-init.service`. We want users to be configuring their ADMIN_PASSWORD declaratively and put it in their config using something like sops-nix/agenix. Running an imperative operation that creates a config, which can easily get lost/moved/deleted in sea of other files (e.g. impermanence) is undesired, especially for a service that is intended to not lose users' files, that'd be ironic :)
>
> Running the `ocis init` command (which is what the `ocis-init.service` does) is required *regardless* of if we provide an option to set the admin password. It generates a bunch of stuff in `/var/lib/ocis/config` (and other subdirs) needed for the system to start properly (see the getting started docs: https://owncloud.dev/ocis/getting-started/).

Pretty sure `ocis init` is not *required*. What is required is the
values of environment variables. We know that because the module test as
it is right now runs without ever invoking `ocis init` :)

> We don't have to run this command, but then user intervention is required before they have a working ocis instance. If we don't run the command, then they would have to run it themselves.
>
> There's plenty of examples in `nixos/modules/*` of similiar type services (just grep for "-init.service" or "init-.*.service"), so I think this is a well established convention.

Yes, but I'd rather the user be made aware of these than have them
hidden away. The list of envvars in module test are the bare minimum it
takes for the services to start. How about, instead of taking the
visibility of everything away, we set the envvars to null, and either
warn or better, error out if the service is enabled without also setting
these envvars?

Let me try to explain this better in next section:

> I do understand your point about imperatively creating a config, but:
>
> 1. All the options that the user can specify with `cfg.settings` override whatever `ocis init` creates in the config file.. so one can still declaratively manage the config with nixos (win!)
> 2. Re sea-of-files/impermanence this is a good point, but the config is created inside the state dir, which has to be persisted anyways (other wise all the files go poof on the next reboot). *And* the created config is in the state dir, so there's no extra runtime path that you have to backup or worry about (like if it wrote something in `/etc`)
> 3. I think it is a much nicer user experience (and also in-line with other nixos modules) that with a config as simple as
>
>     ```
>     services.ocis = {
>         enable = true;
>         settings = {
>           OCIS_INSECURE = "true";
>         };
>     };
>
>     ```
>
>     ... you get a working ocis instance.

Hmm.. This does look nice, but I'm not convinced. As an admin, I want to
be aware of security implications, have good secure defaults AND get to
decide my own risk factor.

Coming to earlier point, if we error out on null variables:

1. we tell user immediately, at the `nixos-rebuild` what they are
missing and what needs to be done. No need to go digging application
docs.

2. It promotes good defaults, by not setting insecure values without
explicit user intervention. We can also have the error docstring point
towards the envvars being included in `environmentFile` instead of
`settings`.

3. Preserves user agency in picking bad/insecure defaults. If I, as a
user want the passwords to be world readable nix store because I protect
it other ways, or that I know all code running on my machine is fully
vetted or that I just want to watch my system burn, the system should *NOT*
be making that decision for me.

4. Provide granularity. As extension of (3.), the user gets to decide
what secrets they want to clutch close to their chests and what they
don't care about for whatever reason.

In nutshell, secure by default + full user agency.

> 4. Nearly everything about nixos modules is hiding imperative actions and instead presenting a nice beautiful declarative interface, so IMO this is doing exactly that.

Disagree. I see nixos modules as 'handling' imperative actions, not
'hiding' them. I still tell the system what I want, system takes care of
'how'. Sounds pedantic, but please entertain me anyway.

For example, my concern with your earlier code snippet is that it tells
me things are insecure, but doesn't tell me why, or how much, or how to
make it secure. The default yaml created by `ocis init` for
fullstack config is big, and the `OCIS_INSECURE` option doesn't tell
user the breadth of choices that are being made for them.

> As for the `cfg.adminPasswordFile` option itself: that option precisely allows the user to use sops/agenix, etc to declaratively specify their admin password (in fact i use sops-nix in my local testing).

I'm guessing this is part of your suggested changes (sorry, haven't been
able to check), but we already have `environmentFile` which serves the same
purpose. Please correct me if wrong.
--

bhankas commented:
Apologies if this came out rant-y, certainly wasn't intentional.

Payas Relekar ***@***.***> writes:

> Casey Link ***@***.***> writes:
>
>> There's plenty of examples in `nixos/modules/*` of similiar type services
>> (just grep for "-init.service" or "init-.*.service"), so I think this is a
>> well established convention.

If you don't mind entertaining me for bit longer, I'm not convinced it
is a good convention, or that we don't have a better alternative for our
situation.

> Yes, but I'd rather the user be made aware of these than have them
> hidden away. The list of envvars in module test are the bare minimum it
> takes for the services to start. How about, instead of taking the
> visibility of everything away, we set the envvars to null, and either
> warn or better, error out if the service is enabled without also setting
> these envvars?

See `services.garage` where the user is forced to pick a package,
we can do something similar for envvars.

>> 1. All the options that the user can specify with `cfg.settings` override
>> whatever `ocis init` creates in the config file.. so one can still
>> declaratively manage the config with nixos (win!)

Yes, but the user does not know what they need to override unless they
dig into documentation of every single service.

>>     ```
>>     services.ocis = {
>>         enable = true;
>>         settings = {
>>           OCIS_INSECURE = "true";
>>         };
>>     };
>>
>>     ```
>>
>>     ... you get a working ocis >

I gave this more thought. A no-frill config that brings everything up is
indeed nice, but I still want to be aware of the full picture and get agency via
NixOS module doc+options, and I don't want to /have to/ dig through
service docs to get started.

How about this?

- We have envvars set to null and error out if they are not set

- We provide a non-standard nixos module options:
  - `services.ocis.auto-generate-secrets` : generate passowrds and jwt_token etc
  - `services.ocis.insecure` : set TLS and *_INSECURE vars to "true"

The semantics of these options can be changed, including whether to have
separate options or combine them.

The idea is,

- User knows what they are getting into. Module option docstring can
  explain which command it runs, what output it generates and where to
  find it.

- User has to explicitly opt into choosing this, and the module can be
  configured to make it exclusive with `serviecs.ocis.settings`. Again,
  this choice is up for further discussion, but I think that promotes
  good balance between secure defaults and user agency.

Just a scratch, perhaps we can explore this more?
--

bhankas commented:
Sorry for the comment barrage, but I just remembered another question.

> Casey Link ***@***.***> writes:
>
>> We don't have to run this command, but then user intervention is required
>> before they have a working ocis instance. If we don't run the command, then
>> they would have to run it themselves.
>>
>> There's plenty of examples in `nixos/modules/*` of similiar type services
>> (just grep for "-init.service" or "init-.*.service"), so I think this is a
>> well established convention.

One more thing to remember is `init-service`, being inherently
imperative, does not play nice with machine migration. It assumes data
(including config generated by init-service) being migrated/copied over
before config.

If the data is migrated after service is already started, then we are on
the hook to ensure service supports (and keep supporting) hot-reload of
config and the copied over stuff has right permissions, something most
backup tools already do, and good admins already practice. But this is
again another thing the user has to keep in mind before migrating.

I'm not a super-power-user, but I have performed multiple device
migrations in past years that I've been using NixOS, and the imperative
`init-service` does not map cleanly to my mental model of how NixOS
config works, where I just copy over config and private key for secret
decryption (I use agenix :)). Everything I ever have to know is in *.nix
files. I know for sure that even if I lose all the data to say
crypto-ransomware (extreme example, please bare with me), my logins and
config keeps working. If something like that does happen, I have bigger
problems, but still one less thing to worry about :)

Not sure this actually supports the argument, but just another thorn in
the haystack.
--

Ramblurr commented:
> Sorry for the comment barrage, but I just remembered another question.

Haha no problem. I'll try to address everything and group related things together.

Let's dive in..

> Pretty sure `ocis init` is not *required*. What is required is the values of environment variables. We know that because the module test as it is right now runs without ever invoking `ocis init` :)

Here is a simple test case to show that the default configuration provided by the module is not sufficient:

```bash

# run from your nixpkgs directory
nix-build -A ocis-bin
mkdir test/
cd test/

export PROXY_HTTP_ADDR="127.0.0.1:9200"
export OCIS_URL="https://localhost:9200"
export OCIS_CONFIG_DIR="${PWD}/config"
export OCIS_BASE_DATA_PATH=${PWD}

../result/bin/ocis server web
````

Result:

```
The jwt_secret has not been set properly in your config for ocis. Make sure your
/home/ramblurr/src/nixpkgs/test/config config contains the proper values (e.g.
by running ocis init or setting it manually in the config/corresponding
environment variable).
```

When I saw this I understood why were adding all those extra `services.ocis.settings` in the test and the example.

Running `ocis init` of course creates the `config/ocis.yaml` with many of those pre-generated.

So when I say "`ocis init` is required", I meant that in the context of a basic
getting-started situation where you specify the minimum amount of required
variables to get a working instance.

I think we are on the same page now about this.

> ... but I still want to be aware of the full picture and get agency via NixOS module doc+options, and I don't want to /have to/ dig through service docs to get started.

I 100% agree. I think?

I expect a NixOS module to allow me to configure every aspect of the software (I
think you agree with that), but I wouldn't expect reading the NixOS options+docs
to give me everything I need to know about being an admin of the service. At
some point I think it is appropriate to punt and expect the admin to go read the
real documentation for the service.

At the same time I think it is a *nice to have* if the nixos module lets me get
up and running quickly without forcing me to go become an expert on the service.
It should provide a balance.

In summary my view:

* the module should help us get started without having to learn the insides and out of how to run the service
* the module MUST let you tweak every knob you need to, especially once you've become an expert in the serivce.

> One more thing to remember is `init-service`, being inherently imperative, does not play nice with machine migration. It assumes data (including config generated by init-service) being migrated/copied over before config.

(snipped rest of that quote)

I *think* I understand what you're saying here. Let me re-phrase it to be sure:

You're saying that

1. the config values generated by `ocis init` in
`config/ocis.yaml` (which consist primarily of secret and passwords) should (or
even *must*) be declaratively defined because if you lose those secrets you
cannot access your data (which I don't think is the case) or ocis will not
correctly service you files.

2. Also, with respect to the "data being migrated/copied over before config",
you're saying that if the user deploys the nixos module, starts it, and then
copies over `/var/lib/ocis` from a backup, the running ocis daemon will get confused.

Re 1: I tested this by running `ocis init` then `serve`, verified I had a
working OCIS. Then I shut it down and removed `ocis.yaml` and generated a new
one with the init command again.

The result was that you could not login and there were several types of errors
printed by the server. So I think it is safe to say that, yes, you should treat
those config vals as *IMPORTANT - DO BACKUP*, which, like you said, a good admin
would do anyways.

Re 2: I don't think it ever makes sense to expect a service to be deployed, then
hot-swap everything underneath it and expect it to just work without at least
some sort of `systemctl restart`

I think it is appropriate to have `ocis init` generate the config (*if the user
explicitly wants*, you've convinced me on that), and especially because it lives
right next to the data itself, it will be included in backups and generating and
managing all those secrets by hand is not necessary if there is a tool that does
it for you.

Anyways this is a bit in the weeds.. back to your suggestion:

> How about this?
..snip..

Yes, I like this approach.

I guess I prefer to treat all those config values that `init` generates as
*data* and not config. Because in most situations you don't need to change them,
and they can be automatically generated (versus *actual config* like hostname,
port, etc)

Of course that data is config, in the sense that you might have to change it
sometime (to rotate secrets in case of a leak *data* is config for example), so
those options should be exposed via the module.

And one negative aspect of my init service (as it stands now) is that while it
won't create a new config yaml if it exists, it will create it even if you
supply all those values as env vars, and in theory the env vars should take
precendence. But in any case it is not clean to have a generated yaml lying
around that isn't actually used. That's confusing.

So yes, let's move forward with your suggestion.

1. Default those settings to `null`, but assert they are `!= null`
2. .. UNLESS something like `services.ocis.generateInitialSecrets = true;`
3. Separate from 1 and 2 is `services.ocis.insecure = true/false;` to manage those variables that let you get running quickly without a cert (which is useful for people just getting started)

The `insecure` flag can always be turned off once the user sets up their certs..
or could never be turned off if they are using a reverse proxy for TLS
termination.

Documentation here is important like you said.

Migrating from `services.ocis.generateInitialSecrets = true;` to `false` isn't
straightforward, the user will have to go find the config file and copy paste
all those values into the right env setting in their `environmentFile`.  The
mapping between the yaml name and the env var name isn't obvious without reading
the docs.

bhankas commented:
Just want to add, great response, thanks!

Combining both replies:

> Yes, I like this approach.
>
> I guess I prefer to treat all those config values that `init` generates as
> *data* and not config. Because in most situations you don't need to change them,
> and they can be automatically generated (versus *actual config* like hostname,
> port, etc)

I see now where you're coming from. The point of contention is that ocis
uses envvars to store both config AND data. But that is for different
mail :p

> So yes, let's move forward with your suggestion.
>
> 1. Default those settings to `null`, but assert they are `!= null`
> 2. .. UNLESS something like `services.ocis.generateInitialSecrets = true;`
> 3. Separate from 1 and 2 is `services.ocis.insecure = true/false;` to manage
> those variables that let you get running quickly without a cert (which is useful
> for people just getting started)

Looks good!

> The `insecure` flag can always be turned off once the user sets up their certs..
> or could never be turned off if they are using a reverse proxy for TLS
> termination.
>
> Documentation here is important like you said.

Yeah, this whole thing needs elaborate docstrings everywhere, something
to note, and we do not want the module to go in stable release without.

> Migrating from `services.ocis.generateInitialSecrets = true;` to `false` isn't
> straightforward, the user will have to go find the config file and copy paste
> all those values into the right env setting in their `environmentFile`.  The
> mapping between the yaml name and the env var name isn't obvious without reading
> the docs.

I'm still trying to think how this can be presented to the user.
Docstrings is one thing, but we want real feedback on what state is
being changed when it is being changed.

e.g. docstring can tell user /before/ adding to config what value they
override. But it gives bad (no) feedback when in 6 months when user
comes back to config for changing something and of course, config
doesn't have docstrings. So now user again has to go to module doc, read
all the options they have enabled (including `generateInitialSecrets`),
infer the yaml file location, read that, and add another variable to
override it. Lots of words to say I see 2 issues:

1. envvar names and YAML values do NOT always map cleanly and
transparently. e.g. OCIS_JWT_SECRET applies to all services as the
documentation says, but OCIS_PASSWORD doesn't.

2. when the user returns to make a change, the `generateInitialSecrets`
option could have already been disabled or even removed. Now the user
has a state they depend on, but not cleanly reflected in config. Since
we are (and can) only asserting limited number of bare minimum envvars,
this creates an information asymmetry in actual state vs what the config
describes.

Considering second point in particular, I think it would be rather
prudent if we scrap envvars entirely, and instead use the `settings`
option to generate YAML instead. This solves both problems immediately:

1. What is in config maps cleanly to yaml. This makes it dead easy to
compare the Nix code to YAML in service documentation. Also allows
de-duplication of common values.

2. If the module had `generateInitialSecrets = true` at any point in
past, the user is clearly using it. If they now try to override
particular values via `settings` there is a conflict, and immediate
feedback to user on what file will be replaced, at which point user can
see what the state of config + data actually is.

(2.) is bit problematic, because there is no gradual override of values,
but all-or-nothing. I think that is okay, because now the user can make
the decision about wanting to replace single value in YAML directly, or
putting entire YAML in config, or deploy it via sops/age. Less
convenient, more explicit. Would like to know what you think on this.

> Not to further complicate things.. but I just went and looked at my own prod
> instances of ocis (I have several, not running on nixos, something I hope to
> change :D).

Had you lead with this, I'd have probably bowed out long ago :p Thanks for
being cool :)

> In order to get ocis to work properly with my IDP (like keycloak) I had to
> supply a
> [`config/proxy.yaml`](https://doc.owncloud.com/ocis/next/deployment/services/s-list/proxy.html#automatic-role-assignments)
> file to set some settings that are *not* available to set as environment
> variables.

Sigh, why do oC folks make these questionable decisions? Nevermind, back
to topic..

> In our current version of the module there is no easy way to drop config a file
> into `/var/lib/ocis/config`.
>
> Some options here:
>
> 1. We move `OCIS_CONFIG_DIR` to `/etc/ocis`
>     * technically I guess this is already possible by overriding `OCIS_CONFIG_DIR` in my settings
>     * this lets the admin (or the module) use `environment.etc` to populate extra config files.
>
> 2. We add something like `services.ocis.configFiles` (`type = types.listOf types.path;`) that copies them for you.
>    * How would the module copy them? It would be some sort of nasty symlink or runtime `install ...` script. Not a fan of this option
>
>What if we need to specify a config file that contains secrets? OCIS
>currently doesn't let us specify the actual path of `proxy.yaml` only the `OCIS_CONFIG_DIR`.

I quite like 2 actually. We can have a service option to specify origin path for proxy.yaml and
Nix can symlink it at runtime. The option docstring can have same
warning about being world readable unless encrypted yadda yadda. I think
that is good enough.

--

Ramblurr commented:
> I think it would be rather prudent if we scrap envvars entirely, and instead use the `settings` option to generate YAML instead.

So.. initially I thought this was a good idea. But by the time I finished writing the reply, and in particularly, working out some examples...I think it results in insanely verbose configs that are a pain to work with.


(here starts the initial reply) Yes, I think this is a good idea. Importantly though since the `ocis.yaml` config file contains secrets it cannot be a file that is stored in the nix store. But nixos already has a couple ways to manage this. Here's some ideas for how to handle it:

1. User supplies entire config yaml using something like `services.ocis.configFile`
    * This bypasses all settings we would set in the module. The admin is 100% responsible for ensuring it has the correct values.
    * The file can by a sops/agenix file that is symlinked into place, or if the admin wants, just a plain text file they plop somewhere.
2. We expose `services.ocis.config` which takes a nixlang datastructure that is serialized to yaml using `lib.generators.toYAML`
    * We do not try to enumerate every possible valid option in this datastructure, rather we just serialize it as is.
    * We can use the [`_secret` convention](https://search.nixos.org/options?query=settings+_secret) combined with `replace-secret` to allow embedding of secrets 

One problem with eschewing env vars is that there are 15 separate secrets that are required to bootstrap an ocis instance, but these secrets need to be placed in the config.yaml file multiple times: https://gist.github.com/Ramblurr/9698ae25913a2370c2188cc1d70d8529

I think this problem is well illustrated with the following examples.

Here are three example usages of the nixos module as we're talking about it now:

### Basic, getting started

```nix
services.ocis = {
  enable = true;
  generateInitialSecrets = true;
  insecure = true;
  config = {
    idm = {
      service_user_passwords = {
        admin_password._secret = "/run/secrets/ocis-admin-pass";
      };          
    };
  };
};
```

### Advanced, admin supplies all config

```nix
services.ocis = {
  enable = true;
  configFile = "/run/secrets/ocis.yaml";
};
```

### Intermediate, supplying config in yaml


```nix
  services.ocis = {
    enable = true;
    config = {
      log = { # putting this here just to show what non-secrets look like
        pretty = true;
        color = true;
        level = "info";
      };

      token_manager = {
        jwt_secret._secret._secret = config.sops.secrets.ocis.jwt_secret.path;
      };
      machine_auth_api_key._secret._secret = config.sops.secrets.ocis.machine_auth_api_key.path;
      system_user_api_key._secret._secret = config.sops.secrets.ocis.system_user_api_key.path;
      transfer_secret._secret._secret = config.sops.secrets.ocis.transfer_secret.path;
      system_user_id._secret._secret = config.sops.secrets.ocis.system_user_id.path;
      admin_user_id._secret._secret = config.sops.secrets.ocis.admin_user_id.path;
      graph = {
        application = {
          id._secret = config.sops.secrets.ocis_graph_application_id.path;
        };
        identity = {
          ldap = {
            bind_password._secret = config.sops.secrets.ocis_graph_identity_ldap_bind_password.path;
          };
        };

        service_account = {
          service_account_id._secret = config.sops.secrets.ocis_service_account_id.path;
          service_account_secret._secret = config.sops.secrets.ocis_service_account_secret.path;
        };
        idp = {
          ldap = {
            bind_password._secret = config.sops.secrets.ocis_graph_idp_ldap_bind_password.path;
          };
        };
        idm = {
          service_user_passwords = {
            admin_password._secret._secret =
              config.sops.secrets.ocis_graph_idm_service_user_admin_password.path;
            idm_password._secret = config.sops.secrets.ocis_graph_idm_service_user_idm_password.path;
            reva_password._secret = config.sops.secrets.ocis_graph_idm_service_user_reva_password.path;
            idp_password._secret = config.sops.secrets.ocis_graph_idm_service_user_idp_password.path;
          };
        };

        proxy = {
          service_account = {
            service_account_id._secret = config.sops.secrets.ocis_service_account_id.path;
            service_account_secret._secret = config.sops.secrets.ocis_service_account_secret.path;
          };
        };

        frontend = {
          service_account = {
            service_account_id._secret = config.sops.secrets.ocis_service_account_id.path;
            service_account_secret._secret = config.sops.secrets.ocis_service_account_secret.path;
          };
        };

        auth_basic = {
          auth_providers = {
            ldap = {
              bind_password._secret = config.sops.secrets.ocis_ldap_bind_password.path;
            };
          };
        };

        users = {
          drivers = {
            ldap = {
              bind_password._secret = config.sops.secrets.ocis_ldap_bind_password.path;
            };
          };
        };

        groups = {
          drivers = {
            ldap = {
              bind_password._secret = config.sops.secrets.ocis_ldap_bind_password.path;
            };
          };
        };

        ocm = {
          service_account = {
            service_account_id._secret = config.sops.secrets.ocis_service_account_id.path;
            service_account_secret._secret = config.sops.secrets.ocis_service_account_secret.path;
          };
        };
        thumbnails = {
          thumbnail = {
            transfer_secret._secret = config.sops.secrets.ocis.thumbnails_transfer_secret.path;
          };
        };
        search = {
          service_account = {
            service_account_id._secret = config.sops.secrets.ocis_service_account_id.path;
            service_account_secret._secret = config.sops.secrets.ocis_service_account_secret.path;
          };
        };

        settings = {
          service_account_ids = [ config.sops.secrets.ocis_service_account_id.path ];
        };
        storage_users = {
          mount_id._secret = config.sops.secrets.ocis.storage_users_mount_id.path;
          service_account = {
            service_account_id._secret = config.sops.secrets.ocis_service_account_id.path;
            service_account_secret._secret = config.sops.secrets.ocis_service_account_secret.path;
          };
        };

        notifications = {
          service_account = {
            service_account_id._secret = config.sops.secrets.ocis_service_account_id.path;
            service_account_secret._secret = config.sops.secrets.ocis_service_account_secret.path;
          };
        };
        gateway = {
          storage_registry = {
            storage_users_mount_id =
              config.sops.secrets.ocis.gateway_storage_registry_storage_users_mount_id.path;
          };
        };

        userlog = {
          service_account = {
            service_account_id._secret = config.sops.secrets.ocis_service_account_id.path;
            service_account_secret._secret = config.sops.secrets.ocis_service_account_secret.path;
          };
        };

        auth_service = {
          service_account = {
            service_account_id._secret = config.sops.secrets.ocis_service_account_id.path;
            service_account_secret._secret = config.sops.secrets.ocis_service_account_secret.path;
          };
        };

        clientlog = {
          service_account = {
            service_account_id._secret = config.sops.secrets.ocis_service_account_id.path;
            service_account_secret._secret = config.sops.secrets.ocis_service_account_secret.path;
          };
        };
      };
    };
  };
```

:nauseated_face: :vomiting_face: 

Why would anyone ever go through that exercise of converting from yaml->nix ? I just did it once for the sake of this example and never want to do it again.

Especially since 99% of people will be starting with values generated from `ocis init` (whether manually ran or via a generator we provide), the process of going from yaml ---(manually)--> nix to just then go again nix ---> yaml is unnecessary imo.

Give a choice of the three examples most sane people will choose the first and second one. Which means all the code required (and the effort to produce and maintain it) to support the third one (nix->yaml w/ secret substitution) will be in vain.

> (2.) is bit problematic, because there is no gradual override of values, but all-or-nothing. 

I think my example shows it *is* a problem?

Even the ocis docs seem to admit that the config yaml format is super verbose, which is why many env vars end up setting multiple yaml options (`OCIS_URL` for example).

I am now leaning to go back to what you had originally :laughing: :

* Just work with env vars, this avoids the yaml explosion
* Do not provide an automatic `services.ocis.generateInitialSecrets` option, this avoids the problem that led us down the road yaml explosion
* ...plus something to handle the `proxy.yaml` case.

I could write a simple script that runs `ocis init` and converts the output into an env var format the user could add to their config as an `environmentFile`, that would be relatively simple and still allow us to have provide an easier getting started path:

Example docs:
Getting started with the nixos OCIS module:

1. Generate your initial secrets

        ocis-nix-init > my-ocis-env

2. Encrypt it with the method of your choice (sops, agenix, etc)

3. Configure the module:

        services.ocis = {
            enable = true;
            environmentFile = config.sops.secrets.ocis-env.path;
        };

What do you think?

> >  2. We add something like `services.ocis.configFiles` (`type = types.listOf types.path;`) that copies them for you.
> 
> I quite like 2 actually. We can have a service option to specify origin path for proxy.yaml and Nix can symlink it at runtime. The option docstring can have same warning about being world readable unless encrypted yadda yadda. I think that is good enough.

Yes I suppose the symlink munging can easily be done with `systemd.tmpfiles.rules` or a bash `ExecStartPre=` script.

danth commented:
Documentation for other NixOS modules tends to simply mention "this should be a path outside the Nix store" rather than pointing users to a specific tool. I think this is important to avoid confusion for new users: if you have a single machine, there is nothing wrong with simply installing the file and making a backup copy by hand.

bhankas commented:
You're right, that YAML soup is awful to look at and nightmare to work with.

Casey Link ***@***.***> writes:

> I am now leaning to go back to what you had originally :laughing: :
>
> * Just work with env vars, this avoids the yaml explosion
> * Do not provide an automatic `services.ocis.generateInitialSecrets` option, this avoids the problem that led us down the road yaml explosion
> * ...plus something to handle the `proxy.yaml` case.
>
> I could write a simple script that runs `ocis init` and converts the output into an env var format the user could add to their config as an `environmentFile`, that would be relatively simple and still allow us to have provide an easier getting started path:

I think this is about time we ask users to think for themselves ;)
Instead of providing any kind of conversion script that we get to
maintain, we ask users to generate initial config directly via
`ocis-init`, and point the module to it. @danth also seems to agree:

> Documentation for other NixOS modules tends to simply mention "this should be a
> path outside the Nix store" rather than pointing users to a specific tool. I
> think this is important to avoid confusion for new users: if you have a single
> machine, there is nothing wrong with simply installing the file and making a
> backup copy by hand.

So, combining with above, let's have 3 options:

1. `configFile` : path to config file, user can generate this via `ocis init` and either have
it directly in store or encrypt it.

2.  `settings` : gets applied to systemd unit as envvars.

3. `environmentFile` : envvars, but in file that user can put wherever.

My reasoning: configFile gets explicitly generated by user, and they can
decide if and how they want to encrypt it. `settings` is the most
convenient option to do anything with. `environmentFile` is the middle
ground between world readable `settings` and black-box `configFile`.
Particularly because of the way OCIS and systemd units give precedence
to values, we have in order of increasing precedence, configFile ->
environmentFile -> settings. I think this provides enough flexibility,
security and convenience.

It does require users to do some work, but heavy lifting will be done by
manual `ocis init` invocation. They have to ensure the config file is
accessible to the service, which can be easily explained by docstring.

WDYT?
--

danth commented:
I would go with `environment` rather than `settings` since it's consistent with the naming of `environmentFile`, and it makes it obvious that it's referring to environment variables rather than the YAML format.

We should also ensure that the error message produced if `configFile` is not set or does not exist is as clear as possible about how to proceed.

bhankas commented:
Okay, I've added a commit with some of the discussed changes, please take a look.

Remaining items:

- adopt test to run `ocis init` first. Little bit of search didn't yield a result so the test is failing.
- adopt @Ramblurr 's suggested test of actually loggin into the service
- Clean up commits

I also forgot about `proxy.yaml`, but I think we can add that to `configDir` docstring. Let me know what you think.

Ramblurr commented:
I've updated my "suggestion" branch:

https://github.com/Ramblurr/nixpkgs/commit/26f526d2fba89967c3ec2ad1187efb728d14c50e

* Added working nixos module test 
* Added a `ocis.md` module documentation
* Added the systemd hardening settings

bhankas commented:
Alright!

@Ramblurr did a bang up job, I adopted your changes pretty much verbatim. Great work! I added you as co-author to module commit, since you already did most of the work there :)

Couldn't quickly find email for @danth , but happy to have it, please let me know if you want.

Also made some tiny changes to docstring and added list of services and their ports to `ocis.md`.

I think this looks good, feel free to take a look.

Ramblurr reviewed (approved):
Tests pass, and I tested it locally and it's working great. And thanks for the co-auther shoutout.

Thanks for all the effort @bhankas! Despite all the discussion we basically ended up where you had gotten us early last week, sorry about that. But I appreciate you entertaining me and I'm very happy with the result.

@MaxwellDupre @adisbladis @danth @SomeoneSerge You've probably tuned this PR out given our back-and-forth (sorry for all that noise), but I believe this PR is ready for merging now.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1570

drupol commented on pkgs/by-name/oc/ocis-bin/package.nix:
```diff
@@ -0,0 +1,56 @@
+  autoPatchelfHook
+}:
+
+stdenv.mkDerivation rec {
```
I think we should use the `finalAttrs` pattern here.

The `finalAttrs` pattern will let you remove the `rec` keyword ([see its implementation in Nix](https://github.com/NixOS/nix/blob/2d1d81114d72ace89ce08cd3bc93f4eb27a2975d/src/libexpr/nixexpr.cc#L121)).

Why is this pattern preferred to `rec` ?

Let's take this simple code example:

```nix
mkDerivation rec {
   foo = 1;
   bar = foo + 1;
}
```

and then `.overrideAttrs(old: { foo = 2; })`, you'll get `{ foo = 2; bar = 2; }` while with `finalAttrs` pattern, it would work correctly because it's a real fixed point.

Let me share a couple of useful links regarding the `finalAttrs` pattern:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

Feel free to reach out if you need some assistance.

bhankas replied:
Done.

drupol commented on pkgs/by-name/oc/ocis-bin/package.nix:
```diff
@@ -0,0 +1,56 @@
+  src = fetchurl {
+    url = "https://github.com/owncloud/ocis/releases/download/v${version}/ocis-${version}-${system}";
+
+    sha256 =
```
Prefer `hash = ...` to `sha256 = ...` Both currently work, but the former will make it easier if the algorithm ever needs to change.

Since there could be different hashing algorithms in use, this make sense.

More info: https://nixos.org/manual/nixpkgs/stable/#fetchurl

> Nixpkgs contributors are currently recommended to use hash.

bhankas replied:
Done

drupol reviewed (approved):
LGTM!
